controllers/grafananotificationtemplate_controller.go

```diff
@@ -0,0 +1,260 @@
+	cl, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return models.NotificationTemplate{}, fmt.Errorf("building grafana client: %w", err)
+	}
+
+	params := provisioning.NewGetTemplateParams().WithName(notificationTemplate.Spec.Name)
+	remote, err := cl.Provisioning.GetTemplateWithParams(params)
+	if err != nil {
+		return models.NotificationTemplate{}, fmt.Errorf("getting notification template: %w", err)
+	}
+	return *remote.Payload, nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) finalize(ctx context.Context, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate) error {
+	r.Log.Info("Finalizing GrafanaNotificationTemplate")
+
+	instances, err := r.GetMatchingInstances(ctx, notificationTemplate, r.Client)
+	if err != nil {
+		return fmt.Errorf("fetching instances: %w", err)
+	}
+	for _, i := range instances {
+		instance := i
+		if err := r.removeFromInstance(ctx, &instance, notificationTemplate); err != nil {
+			return fmt.Errorf("removing contact point from instance: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) removeFromInstance(ctx context.Context, instance *grafanav1beta1.Grafana, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate) error {
+	cl, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return fmt.Errorf("building grafana client: %w", err)
+	}
+
+	_, err = r.getNotificationTemplateByName(ctx, instance, notificationTemplate)
+	if err != nil {
+		return fmt.Errorf("getting notification template by name: %w", err)
+	}
+	_, err = cl.Provisioning.DeleteTemplate(notificationTemplate.Spec.Name) //nolint:errcheck
+	if err != nil {
+		return fmt.Errorf("deleting notification template: %w", err)
+	}
+
+	return nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) GetMatchingInstances(ctx context.Context, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate, k8sClient client.Client) ([]grafanav1beta1.Grafana, error) {
+	instances, err := GetMatchingInstances(ctx, k8sClient, notificationTemplate.Spec.InstanceSelector)
+	if err != nil || len(instances.Items) == 0 {
+		return nil, err
+	}
+	if notificationTemplate.Spec.AllowCrossNamespaceImport != nil && *notificationTemplate.Spec.AllowCrossNamespaceImport {
+		return instances.Items, nil
+	}
+	items := []grafanav1beta1.Grafana{}
+	for _, i := range instances.Items {
+		if i.Namespace == notificationTemplate.Namespace {
+			items = append(items, i)
+		}
+	}
+
+	return items, err
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *GrafanaNotificationTemplateReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&grafanav1beta1.GrafanaNotificationTemplate{}).
+		Complete(r)
```

Comment:
Currently, only a few controllers ignore `Status` updates.
To save a future PR, could you add the filter here? :)
```suggestion
		For(&grafanav1beta1.GrafanaNotificationTemplate{}).
		WithEventFilter(ignoreStatusUpdates()).
		Complete(r)
```